nav2_controller/src/controller_server.cpp

```diff
@@ -451,17 +451,25 @@ void ControllerServer::computeControl()
 
     last_valid_cmd_time_ = now();
     rclcpp::WallRate loop_rate(controller_frequency_);
+    bool cancelling = false;
     while (rclcpp::ok()) {
       if (action_server_ == nullptr || !action_server_->is_server_active()) {
         RCLCPP_DEBUG(get_logger(), "Action server unavailable or inactive. Stopping.");
         return;
       }
 
       if (action_server_->is_cancel_requested()) {
-        RCLCPP_INFO(get_logger(), "Goal was canceled. Stopping the robot.");
-        action_server_->terminate_all();
-        publishZeroVelocity();
-        return;
+        if (controllers_[current_controller_]->cancel()) {
+          RCLCPP_INFO(get_logger(), "Cancellation was successful. Stopping the robot.");
+          action_server_->terminate_all();
+          publishZeroVelocity();
+          return;
+        } else {
+          if (!cancelling) {
+            RCLCPP_INFO(get_logger(), "Goal was canceled. Trying to stop the robot.");
```

Comment:
If you make this a throttle log, then you don't need to have this `if(not var) {var = !var}` logic and put this all in 1 line. Having a regular 1s reminder about what's happening still would be useful I think too